hbase-server/src/test/java/org/apache/hadoop/hbase/ClearUserNamespacesAndTablesRule.java

```diff
@@ -0,0 +1,136 @@
+    return CompletableFuture.runAsync(() -> logger.info("listing tables"))
+      .thenCompose(_void -> admin.listTableNames(false))
+      .thenApply(tableNames -> {
+        final StringJoiner joiner = new StringJoiner(", ", "[", "]");
+        tableNames.stream().map(TableName::getNameAsString).forEach(joiner::add);
+        logger.info("found existing tables {}", joiner.toString());
+        return tableNames;
+      });
+  }
+
+  private CompletableFuture<Boolean> isTableEnabled(final TableName tableName) {
+    return admin.isTableEnabled(tableName)
+      .thenApply(isEnabled -> {
+        logger.info("table {} is enabled.", tableName);
+        return isEnabled;
+      });
+  }
+
+  private CompletableFuture<Void> disableIfEnabled(final TableName tableName) {
+    return isTableEnabled(tableName)
+      .thenCompose(isEnabled -> {
+        if (isEnabled) { return disableTable(tableName); }
+        return CompletableFuture.completedFuture(null);
+      });
+  }
+
+  private CompletableFuture<Void> disableTable(final TableName tableName) {
+    return CompletableFuture.runAsync(() -> logger.info("disabling enabled table {}", tableName))
+      .thenCompose(_void -> admin.disableTable(tableName));
+  }
+
+  private CompletableFuture<Void> deleteTable(final TableName tableName) {
+    return CompletableFuture.runAsync(() -> logger.info("deleting disabled table {}", tableName))
+      .thenCompose(_void -> admin.deleteTable(tableName));
+  }
+
+  private CompletableFuture<List<String>> listUserNamespaces() {
+    return CompletableFuture.runAsync(() -> logger.info("listing namespaces"))
+      .thenCompose(_void -> admin.listNamespaceDescriptors())
+      .thenApply(namespaceDescriptors -> {
+        final StringJoiner joiner = new StringJoiner(", ", "[", "]");
+        final List<String> names = namespaceDescriptors.stream()
+          .map(NamespaceDescriptor::getName)
+          .peek(joiner::add)
+          .collect(Collectors.toList());
+        logger.info("found existing namespaces {}", joiner.toString());
+        return names;
+      })
+      .thenApply(namespaces -> namespaces.stream()
+        .filter(namespace -> !Objects.equals(
+          namespace, NamespaceDescriptor.SYSTEM_NAMESPACE.getName()))
+        .filter(namespace -> !Objects.equals(
+          namespace, NamespaceDescriptor.DEFAULT_NAMESPACE.getName()))
+        .collect(Collectors.toList()));
+  }
+
+  private CompletableFuture<Void> deleteNamespace(final String namespace) {
+    return CompletableFuture.runAsync(() -> logger.info("deleting namespace {}", namespace))
+      .thenCompose(_void -> admin.deleteNamespace(namespace));
+  }
+
+  private CompletableFuture<Void> deleteUserNamespaces() {
+    return listUserNamespaces()
+      .thenCompose(namespaces -> CompletableFuture.allOf(
+        namespaces.stream()
+          .map(this::deleteNamespace)
+          .toArray(CompletableFuture[]::new)));
+  }
+}
```

Comment:
One thought, why cleanup after a test? What if it takes a bunch of work cleaning up, more than just start fresh?

Reply:
I'm writing tests over the content of meta. So I can
1. clean up after each test method; or
1. tear down and spin up a new mini cluster for each test method; or
1. create a fixed set of tables `@BeforeClass` and write all test methods around those tables.

I chose the first option because (2) would make for a really slow test run and (3) is very inflexible in terms of how each test method is written, and renders all other test methods brittle to tweaks of any one of them.